### PR TITLE
Introduce new managed transaction methods and deprecate existing ones

### PR DESCRIPTION
**Pull request by injectives**
This update introduces new methods for managed transactions, also known as transaction functions. Furthermore, it deprecates the existing methods.

The new methods are:
- `executeRead` (synchronous and reactive), `executeReadAsync` (asynchronous)
- `executeWrite` (synchronous and reactive), `executeWriteAsync` (asynchronous)
- `executeWriteWithoutResult` (synchronous)

There are overloaded options that take transaction configuration.

The new methods do not permit explicit management (commit, rollback, close) of managed transaction. The driver will attempt committing the transaction as long as the provided unit of work completes successfully.

### CODE REVIEW

**Review by gjmwoods — APPROVED** [review]
A nice simplification of the tx function API

File: driver/clirr-ignored-differences.xml [hunk 1]
```diff
@@ -122,4 +122,150 @@
         <method>org.reactivestreams.Publisher isOpen()</method>
     </difference>
 
+    <difference>
```
Comment (michael-simons) [inline driver/clirr-ignored-differences.xml:125]: Doesn't clirr support "this is a big change, we bump major anyway, please leave me alone?"

File: driver/src/main/java/org/neo4j/driver/Session.java [hunk 2]
```diff
@@ -87,67 +88,169 @@
...
      * @return a result as returned by the given unit of work.
+     * @deprecated superseded by {@link #executeRead(TransactionCallback)}.
      */
+    @Deprecated
     <T> T readTransaction( TransactionWork<T> work );
 
     /**
-     * Execute a unit of work in a managed {@link AccessMode#READ read} transaction
-     * with the specified {@link TransactionConfig configuration}.
+     * Execute a unit of work as a single managed transaction with {@link AccessMode#READ read} access mode and retry behaviour.
```
Comment (michael-simons) [inline driver/src/main/java/org/neo4j/driver/Session.java]:
```suggestion
     * Execute a unit of work as a single, managed transaction with {@link AccessMode#READ read} access mode and retry behaviour. The transaction allows for one or more statements to be run.
```

Something similar on the other variants (the info about one or more statements).

File: driver/src/main/java/org/neo4j/driver/Session.java [hunk 3]
```diff
@@ -87,67 +88,169 @@
...
      * <p>
-     * This transaction will automatically be committed unless an exception is
-     * thrown during query execution or by the user code.
+     * The driver will attempt committing the transaction when the provided unit of work completes successfully. A user initiated failure of the unit of work
```
Comment (michael-simons) [inline driver/src/main/java/org/neo4j/driver/Session.java]:
```suggestion
     * The driver will attempt committing the transaction when the provided unit of work completes successfully. Any runtime exception thrown by the unit of work
```

File: driver/src/main/java/org/neo4j/driver/Session.java [hunk 4]
```diff
@@ -87,67 +88,169 @@
...
+     * will result in rollback attempt.
```
Comment (michael-simons) [inline driver/src/main/java/org/neo4j/driver/Session.java]:
```suggestion
     * will result in a rollback attempt.
```

File: driver/src/main/java/org/neo4j/driver/Session.java [hunk 5]
```diff
@@ -87,67 +88,169 @@
...
      * <p>
-     * Managed transactions should not generally be explicitly committed (via
-     * {@link Transaction#commit()}).
+     * The provided unit of work should not return {@link Result} object.
```
Comment (michael-simons) [inline driver/src/main/java/org/neo4j/driver/Session.java]:
```suggestion
     * The provided unit of work should not return {@link Result} object as it won't be valid outside the scope of the transaction.
```

File: driver/src/main/java/org/neo4j/driver/Session.java [hunk 6]
```diff
@@ -87,67 +88,169 @@
...
+    {
+        return executeRead( callback, TransactionConfig.empty() );
+    }
+
+    /**
+     * Execute a unit of work in a managed {@link AccessMode#READ read} transaction with the specified {@link TransactionConfig configuration}.
+     * <p>
+     * This transaction will automatically be committed unless an exception is thrown during query execution or by the user code.
+     * <p>
+     * Managed transactions should not generally be explicitly committed (via {@link Transaction#commit()}).
```
Comment (michael-simons) [inline driver/src/main/java/org/neo4j/driver/Session.java]:
```suggestion
     * Managed transactions should not be explicitly committed (via {@link Transaction#commit()}).
```

File: driver/src/main/java/org/neo4j/driver/Session.java [hunk 7]
```diff
@@ -87,67 +88,169 @@
...
     /**
      * Execute a unit of work in a managed {@link AccessMode#WRITE write} transaction.
      * <p>
-     * This transaction will automatically be committed unless an exception is
-     * thrown during query execution or by the user code.
+     * This transaction will automatically be committed unless an exception is thrown during query execution or by the user code.
      * <p>
-     * Managed transactions should not generally be explicitly committed (via
-     * {@link Transaction#commit()}).
+     * Managed transactions should not generally be explicitly committed (via {@link Transaction#commit()}).
```
Comment (michael-simons) [inline driver/src/main/java/org/neo4j/driver/Session.java]:
```suggestion
     * Managed transactions should not be explicitly committed (via {@link Transaction#commit()}).
```

File: driver/src/main/java/org/neo4j/driver/Session.java [hunk 8]
```diff
@@ -87,67 +88,169 @@
...
      * @return a result as returned by the given unit of work.
+     * @deprecated superseded by {@link #executeWrite(TransactionCallback)}.
      */
+    @Deprecated
     <T> T writeTransaction( TransactionWork<T> work );
 
     /**
-     * Execute a unit of work in a managed {@link AccessMode#WRITE write} transaction
-     * with the specified {@link TransactionConfig configuration}.
+     * Execute a unit of work as a single managed transaction with {@link AccessMode#WRITE write} access mode and retry behaviour.
```
Comment (michael-simons) [inline driver/src/main/java/org/neo4j/driver/Session.java]:
```suggestion
     * Execute a unit of work as a single, managed transaction with {@link AccessMode#WRITE write} access mode and retry behaviour.
```

File: driver/src/main/java/org/neo4j/driver/Session.java [hunk 9]
```diff
@@ -87,67 +88,169 @@
...
      * <p>
-     * This transaction will automatically be committed unless an exception is
-     * thrown during query execution or by the user code.
+     * The driver will attempt committing the transaction when the provided unit of work completes successfully. A user initiated failure of the unit of work
+     * will result in rollback attempt.
```
Comment (michael-simons) [inline driver/src/main/java/org/neo4j/driver/Session.java]:
```suggestion
     * The driver will attempt committing the transaction when the provided unit of work completes successfully. Any runtime exception thrown by of the unit of work
     * will result in a rollback attempt.
```

File: driver/src/main/java/org/neo4j/driver/Session.java [hunk 10]
```diff
@@ -87,67 +88,169 @@
...
      * <p>
-     * Managed transactions should not generally be explicitly committed (via
-     * {@link Transaction#commit()}).
+     * The provided unit of work should not return {@link Result} object.
```
Comment (michael-simons) [inline driver/src/main/java/org/neo4j/driver/Session.java]:
```suggestion
     * The provided unit of work should not return {@link Result} object as it won't be valid outside the scope of the transaction.
```

File: driver/src/main/java/org/neo4j/driver/Session.java [hunk 11]
```diff
@@ -87,67 +88,169 @@
...
+    default <T> T executeWrite( TransactionCallback<T> callback )
+    {
+        return executeWrite( callback, TransactionConfig.empty() );
+    }
+
+    /**
+     * Execute a unit of work as a single managed transaction with {@link AccessMode#WRITE write} access mode and retry behaviour.
+     * <p>
+     * The driver will attempt committing the transaction when the provided unit of work completes successfully. A user initiated failure of the unit of work
+     * will result in rollback attempt.
```
Comment (michael-simons) [inline driver/src/main/java/org/neo4j/driver/Session.java]:
```suggestion
     * The driver will attempt committing the transaction when the provided unit of work completes successfully. Any runtime exception thrown by of the unit of work
     * will result in a rollback attempt.
```

File: driver/src/main/java/org/neo4j/driver/Session.java [hunk 12]
```diff
@@ -87,67 +88,169 @@
...
+     * <p>
+     * The provided unit of work should not return {@link Result} object.
```
Comment (michael-simons) [inline driver/src/main/java/org/neo4j/driver/Session.java]:
```suggestion
     * The provided unit of work should not return {@link Result} object as it won't be valid outside the scope of the transaction.
```

File: driver/src/main/java/org/neo4j/driver/Session.java [hunk 13]
```diff
@@ -87,67 +88,169 @@
...
+                          return null;
+                      } );
+    }
+
+    /**
+     * Execute a unit of work in a managed {@link AccessMode#WRITE write} transaction with the specified {@link TransactionConfig configuration}.
+     * <p>
+     * This transaction will automatically be committed unless an exception is thrown during query execution or by the user code.
+     * <p>
+     * Managed transactions should not generally be explicitly committed (via {@link Transaction#commit()}).
```
Comment (michael-simons) [inline driver/src/main/java/org/neo4j/driver/Session.java]:
```suggestion
     * Managed transactions should not be explicitly committed (via {@link Transaction#commit()}).
```

File: driver/src/main/java/org/neo4j/driver/Session.java [hunk 14]
```diff
@@ -87,67 +88,169 @@
...
 
     /**
-     * Run a query in a managed auto-commit transaction with the specified
-     * {@link TransactionConfig configuration}, and return a result stream.
+     * Execute a unit of work as a single managed transaction with {@link AccessMode#WRITE write} access mode and retry behaviour.
+     * <p>
+     * The driver will attempt committing the transaction when the provided unit of work completes successfully. A user initiated failure of the unit of work
+     * will result in rollback attempt.
+     * <p>
+     * The provided unit of work should not return {@link Result} object.
```
Comment (michael-simons) [inline driver/src/main/java/org/neo4j/driver/Session.java]:
```suggestion
     * The provided unit of work should not return {@link Result} object as it won't be valid outside the scope of the transaction.
```

File: driver/src/main/java/org/neo4j/driver/Session.java [hunk 15]
```diff
@@ -87,67 +88,169 @@
...
+    @Deprecated
     <T> T writeTransaction( TransactionWork<T> work, TransactionConfig config );
 
     /**
-     * Run a query in a managed auto-commit transaction with the specified
-     * {@link TransactionConfig configuration}, and return a result stream.
+     * Execute a unit of work as a single managed transaction with {@link AccessMode#WRITE write} access mode and retry behaviour.
+     * <p>
+     * The driver will attempt committing the transaction when the provided unit of work completes successfully. A user initiated failure of the unit of work
+     * will result in rollback attempt.
```
Comment (michael-simons) [inline driver/src/main/java/org/neo4j/driver/Session.java]:
```suggestion
     * The driver will attempt committing the transaction when the provided unit of work completes successfully. Any runtime exception thrown by of the unit of work
     * will result in a rollback attempt.
```

File: driver/src/main/java/org/neo4j/driver/Session.java [hunk 16]
```diff
@@ -87,67 +88,169 @@
...
+     * @param <T>      the return type of the given unit of work.
+     * @return a result as returned by the given unit of work.
+     */
+    <T> T executeWrite( TransactionCallback<T> callback, TransactionConfig config );
+
+    /**
+     * Execute a unit of work as a single managed transaction with {@link AccessMode#WRITE write} access mode and retry behaviour.
+     * <p>
+     * The driver will attempt committing the transaction when the provided unit of work completes successfully. A user initiated failure of the unit of work
+     * will result in rollback attempt.
```
Comment (michael-simons) [inline driver/src/main/java/org/neo4j/driver/Session.java]:
```suggestion
     * The driver will attempt committing the transaction when the provided unit of work completes successfully. Any runtime exception thrown by of the unit of work
     * will result in a rollback attempt.
```

File: driver/src/main/java/org/neo4j/driver/Session.java [hunk 17]
```diff
@@ -87,67 +88,169 @@
...
+     * <p>
+     * The provided unit of work should not return {@link Result} object.
```
Comment (michael-simons) [inline driver/src/main/java/org/neo4j/driver/Session.java]:
```suggestion
     * The provided unit of work should not return {@link Result} object as it won't be valid outside the scope of the transaction.
```

File: driver/src/main/java/org/neo4j/driver/SimpleQueryRunner.java [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.neo4j.driver;
+
+import java.util.Map;
+
+/**
```
Comment (michael-simons) [inline driver/src/main/java/org/neo4j/driver/SimpleQueryRunner.java:23]: Great comment!!!

File: driver/src/main/java/org/neo4j/driver/TransactionCallback.java [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+
+/**
+ * Callback that executes operations against a given {@link TransactionContext}.
+ *
+ * @param <T> the return type of this work.
+ */
+public interface TransactionCallback<T>
+{
+    /**
+     * Executes all given operations against the same transaction context.
```
Comment (michael-simons) [inline driver/src/main/java/org/neo4j/driver/TransactionCallback.java]:
```suggestion
     * Executes all given operations in the same transaction context.
```

**Review by michael-simons — CHANGES_REQUESTED** [review]
This is a great change to the surface of the driver, thank you for that!

I went in detail through all the JavaDoc. I appreciate the information about the semantics, it should go in a doc, too.

I annotated the blocking session with a couple of suggestions. Please apply them to the standard session and do the same for async and reactive.

Thank you.

File: driver/src/main/java/org/neo4j/driver/Session.java (hunk 2 above)
Comment (injectives) [reply driver/src/main/java/org/neo4j/driver/Session.java]: Thanks, updated.

File: driver/src/main/java/org/neo4j/driver/Session.java (hunk 4 above)
Comment (injectives) [reply driver/src/main/java/org/neo4j/driver/Session.java]: Thanks, updated.

File: driver/src/main/java/org/neo4j/driver/Session.java (hunk 5 above)
Comment (injectives) [reply driver/src/main/java/org/neo4j/driver/Session.java]: Thanks, updated.

File: driver/src/main/java/org/neo4j/driver/Session.java (hunk 17 above)
Comment (injectives) [reply driver/src/main/java/org/neo4j/driver/Session.java]: Thanks, updated.

File: driver/src/main/java/org/neo4j/driver/Session.java (hunk 6 above)
Comment (injectives) [reply driver/src/main/java/org/neo4j/driver/Session.java]: Thanks, updated.

File: driver/src/main/java/org/neo4j/driver/Session.java (hunk 7 above)
Comment (injectives) [reply driver/src/main/java/org/neo4j/driver/Session.java]: Thanks, updated.

File: driver/src/main/java/org/neo4j/driver/Session.java (hunk 8 above)
Comment (injectives) [reply driver/src/main/java/org/neo4j/driver/Session.java]: Thanks, updated.

File: driver/src/main/java/org/neo4j/driver/Session.java (hunk 10 above)
Comment (injectives) [reply driver/src/main/java/org/neo4j/driver/Session.java]: Thanks, updated.

File: driver/src/main/java/org/neo4j/driver/Session.java (hunk 12 above)
Comment (injectives) [reply driver/src/main/java/org/neo4j/driver/Session.java]: Thanks, updated.

File: driver/src/main/java/org/neo4j/driver/Session.java (hunk 13 above)
Comment (injectives) [reply driver/src/main/java/org/neo4j/driver/Session.java]: Thanks, updated.

File: driver/src/main/java/org/neo4j/driver/Session.java (hunk 14 above)
Comment (injectives) [reply driver/src/main/java/org/neo4j/driver/Session.java]: Thanks, updated.

File: driver/src/main/java/org/neo4j/driver/TransactionCallback.java (hunk 19 above)
Comment (injectives) [reply driver/src/main/java/org/neo4j/driver/TransactionCallback.java]: Thanks, updated.

File: driver/clirr-ignored-differences.xml (hunk 1 above)
Comment (injectives) [reply driver/clirr-ignored-differences.xml:125]: I might explore later. I find it useful for keeping track of these updates and it serves as a changelog at times.

File: driver/src/main/java/org/neo4j/driver/Session.java (hunk 3 above)
Comment (injectives) [reply driver/src/main/java/org/neo4j/driver/Session.java]: Thanks. I have made this part a bit more detailed.

File: driver/src/main/java/org/neo4j/driver/Session.java (hunk 15 above)
Comment (injectives) [reply driver/src/main/java/org/neo4j/driver/Session.java]: Thanks. I have made this part a bit more detailed.

File: driver/src/main/java/org/neo4j/driver/Session.java (hunk 16 above)
Comment (injectives) [reply driver/src/main/java/org/neo4j/driver/Session.java]: Thanks. I have made this part a bit more detailed.

File: driver/src/main/java/org/neo4j/driver/Session.java (hunk 9 above)
Comment (injectives) [reply driver/src/main/java/org/neo4j/driver/Session.java]: Thanks. I have made this part a bit more detailed.

File: driver/src/main/java/org/neo4j/driver/Session.java (hunk 11 above)
Comment (injectives) [reply driver/src/main/java/org/neo4j/driver/Session.java]: Thanks. I have made this part a bit more detailed.

**Review by michael-simons — APPROVED** [review]

**Pull request merged**